### ISSUE TITLE
When players drop a token on the map, naming preferences are not applied

### ISSUE THREAD

**Issue by Azhrei** [RPTools/maptool#2173]
**Describe the bug**
When a player drops an `.rptok` token on the map, *Preferences* settings for preventing duplicate token names are not applied.

**To Reproduce**
1. Open a new campaign.
2. Start a server; select **Player** instead of **GM** (saves a few steps).
3. Drag an image to the map from the **Default** *Resource Library*.
4. Right-click and save the token somewhere in the *Resource Library*.
5. Navigate to that location in the Resource Library and drag the same token to the map.
6. An error appears in chat that duplicate token names are not allowed, regardless of *Preferences* settings.

**Expected behavior**
Preferences for MapTool include three features for controlling how tokens are named when dragged to the map from the Resource Library.

+ Use the filename as the token name (or the word `Creature`).
+ Use the token name, GM name, or label field as the token name (for use with `.rptok` files).
+ Add a numeric suffix to make the name unique (either randomly generated or sequential).

These preferences should apply to players as well as the GM.  Otherwise, players cannot prepare a token to represent a summoned creature or spell effect and drag multiple of them to the map.

**MapTool Info**
- Version: 1.7.0

**Desktop (please complete the following information):**
 - OS: macOS (applies to all?)
 - Version 10.15.6 (latest)

**Additional context**
This applies specifically to `.rptok` files.  Normal PNG and JPG images are affected correctly by the *Preferences* settings.

**Comment by Phergus** [comment]
Because Players drop PCs and when you saved it the token was saved as a PC.  Change it to an NPC first and then save it.  NPC tokens dropped in by Players will get the naming preferences applied.  (They also get turned into PCs but that's another issue.)

**Comment by Azhrei** [comment]
> Because Players drop PCs and when you saved it the token was saved as a PC. Change it to an NPC first and then save it. NPC tokens dropped in by Players will get the naming preferences applied. (They also get turned into PCs but that's another issue.)

It was saved as GM with the NPC flag set, but you're correct that they are specifically set to PC when dropped by a player.

However, attempting to drop an `.rptok` with the NPC flag set doesn't change the result (it fails regardless).

It looks like the code is specifically checking to ensure the existing token is owned by the same player and failing if it does not.

        // For Players, check to see if the name is already in use. If it is already in use, make
        // sure the current Player owns the token being duplicated (to avoid subtle ways of
        // manipulating someone else's token!).
        Token tokenNameUsed = zone.getTokenByName(token.getName());
        if (tokenNameUsed != null) {
          if (!AppUtil.playerOwns(tokenNameUsed)) {
            failedPaste.add(token.getName());
            continue;
          }
          String newName = MapToolUtil.nextTokenId(zone, token, tokenNameUsed != null);
          token.setName(newName);
        }

Instead of failing, it could just rename the new token being dropped.  Any ideas why that might not have been done in the first place?  (This is ZoneRenderer.java:4645 or so.)

**Comment by Azhrei** [comment]
The GM creating it was part of the problem.  The token did not have the **OwnedByAll** flag set, and the player didn't own it either, hence it failed.  If I had saved the token _as the player_, it would've worked fine because the player would've been an owner and the IF statement would've fallen through.

Regardless, whether the player is or isn't the owner, the _Preferences_ should still be respected, right?

**Comment by Phergus** [comment]
> It was saved as GM with the NPC flag set, but you're correct that they are specifically set to PC when dropped by a player.

That's not the steps you posted.

> Start a server; **select Player instead of GM** (saves a few steps).
> Drag an image to the map from the Default Resource Library.
> Right-click and save the token somewhere in the Resource Library.

If I set the token to NPC before saving it, when they are dropped in by the Player acting as server they follow the naming convention.

**Comment by Azhrei** [comment]
Yep, you're right.  Ugh.  I guess I'll go back and edit the original post.  Thanks.

**Comment by Phergus** [comment]
The code should probably respect the saved PC/NPC state even when a token is dropped by a Player.

**Comment by Azhrei** [comment]
> The code should probably respect the saved PC/NPC state even when a token is dropped by a Player.

Yes, agreed.

Players dropping tokens should perhaps clear all of the VBL information; that tab of the token editor is grayed out for tokens that are player-editable.  I can't think of a reason why a GM might want to allow a player to drop tokens with VBL on them.  Also, the terrain modifiers on the **Config** tab are _not_ grayed out, but that's probably okay because it allows for spell effects that make such changes.
